gquant/dataframe_flow/_node_flow.py

```diff
@@ -138,61 +153,155 @@ def get_pout(df_out):
 
         return output_df
 
+    def validate_connected_ports(self):
+        self_nodetype = _get_nodetype(self)
+        nodetype_names = [inodet.__name__ for inodet in self_nodetype]
+        if 'OutputCollector' in nodetype_names:
+            # Don't validate for OutputCollector
+            return
+
+        msgfmt = '"{task}":"{nodetype}" {inout} port "{ioport}" {inout} port '\
+            'type(s) "{ioport_types}"'
+
+        iports_connected = self.get_connected_inports()
+        iports_spec = self._get_input_ports(full_port_spec=True)
+        for iport in iports_connected.keys():
+            iport_spec = iports_spec[iport]
+            iport_type = iport_spec[PortsSpecSchema.port_type]
+            iport_types = [iport_type] \
+                if not isinstance(iport_type, Iterable) else iport_type
+
+            for ient in self.inputs:
+                # find input node edge entry with to_port, from_port, from_node
+                if not iport == ient['to_port']:
+                    continue
+                ientnode = ient
+                break
+            else:
+                intask = self._task_obj[TaskSpecSchema.inputs][iport]
+                # this should never happen
+                raise LookupError(
+                    'Task "{}" not connected to "{}.{}". Add task spec to '
+                    'TaskGraph.'.format(intask, self.uid, iport))
+
+            from_node = ientnode['from_node']
+            oport = ientnode['from_port']
+            oports_spec = from_node._get_output_ports(full_port_spec=True)
+            oport_spec = oports_spec[oport]
+            oport_type = oport_spec[PortsSpecSchema.port_type]
+            oport_types = [oport_type] \
+                if not isinstance(oport_type, Iterable) else oport_type
+
+            for iptype in iport_types:
+                if issubclass(iptype, tuple(oport_types)):
```

Comment:
This part should be careful. The logic says the input port(child node) type should be a subclass of the output port (parent node). It seems the NeMo Neural types has the opposite rule, maybe I am wrong.

Reply:
You are right. The port type that a consuming node has should match the interface of the incoming port type.
```
class A: 
    pass 

class B(A): 
    pass

class C(A): 
    pass

```
I see your logic. It should be like this:
```
            for optype in oport_types:
                if issubclass(optype , tuple(iport_types)):
```

This makes more sense:
```
iport_types = [C]
oport_types = [A, B]

In [272]: issubclass(A, (C,))                                                                                                                                                                                      
Out[272]: False

In [273]: issubclass(B, (C,))                                                                                                                                                                                      
Out[273]: False

```
Whereas if the inport has the interface of oports it would produce true:
```
iport_types = [A]
oport_types = [B, C]

In [274]: issubclass(B, (A,))                                                                                                                                                                                      
Out[274]: True

In [275]: issubclass(C, (A,))                                                                                                                                                                                      
Out[275]: True

```

I will fix the above.

I'm not sure what the implications are for the NeMo Neural types are. Those Neural types seem obfuscated to me. Aren't they container types and I have to know the details of what a Neural type actually hold to reason about it? Is the columns setup logic sufficient for that?

Reply:
NeMo validation is implemented as the following in the `validator.ts`:
```ts
function validNmType(required: any, outputs: any): boolean {
  //first check types

  const reqElement = required['element'];
  const outElement = outputs['element'];
  if (
    outElement['types'][0] !== 'VoidType' &&
    reqElement['types'][0] !== 'VoidType'
  ) {
    if (
      outElement['types'].findIndex(
        (d: string) => d === reqElement['types'][0]
      ) < 0
    ) {
      // req type should be more specific,
      // out type should be generic
      // first out type should be the parent type of required
      return false;
    }
    if (outElement['fields'] !== reqElement['fields']) {
      return false;
    }
    if (outElement['parameters'] !== reqElement['parameters']) {
      return false;
    }
  }

```
reqElements types is a list of ['class A', 'parent class', 'grandparent class', ...] for input port
outElements types is a list of ['class B', 'parent class', 'grandparent class', ...] for output port
It is only valid if 'classA' ins in the outElements type list, so the output port type has to be a subclass of input port, i.e. class B is subclass of class A. I noticed that my comments is wrong, could you help to fix it too? It should be 
```ts
      // req type should be generic,
      // out type should be specific, i.e. subclass of req
      // first required element type should be the parent type of the output element
```

Reply:
This is a good example and motivation to extract out and refactor the columns validation logic. This kind of logic might be better in backend.

I squashed/pushed with the updates. If you cloned by branch you might need to reclone. I fixed the sublcass comparison order for port types. I'm running through the notebooks checking that nothing broke.

Reply:
Agree. Let's fix it in the future to have the plugin system.